README.md
# phlib/litmus-subjectline-preview

[![Build Status](https://img.shields.io/travis/com/phlib/litmus-subjectline-preview/main.svg)](https://travis-ci.com/phlib/litmus-subjectline-preview)
[![Codecov](https://img.shields.io/codecov/c/github/phlib/litmus-subjectline-preview.svg)](https://codecov.io/gh/phlib/litmus-subjectline-preview)
[![Latest Stable Version](https://img.shields.io/packagist/v/phlib/litmus-subjectline-preview.svg)](https://packagist.org/packages/phlib/litmus-subjectline-preview)
[![Total Downloads](https://img.shields.io/packagist/dt/phlib/litmus-subjectline-preview.svg)](https://packagist.org/packages/phlib/litmus-subjectline-preview)
![Licence](https://img.shields.io/github/license/phlib/litmus-subjectline-preview.svg)

SDK to generate Litmus subject line preview images.

**April 2021**: [Phlib](https://github.com/phlib) have taken over the
maintenance of [`yzalis/litmus`](https://github.com/yzalis/litmus)
(original [Packagist](https://packagist.org/packages/yzalis/litmus))
from [blaugueux](https://github.com/blaugueux), now available as
[`phlib/litmus-reseller-sdk`](https://github.com/phlib/litmus-reseller-sdk).
Huge thanks for trusting us to continue his work! \
This package has been created to split the _subject line preview_ elements.

## Install

Via Composer

```sh
$ composer require phlib/litmus-subjectline-preview
```

## Usage

Check [example.php](example/example.php)

## License

This package is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [3.0.0] - 2021-05-23

### Added
- Add specific support for PHP v7 and v8 (`^7.3 | ^8.0`).
- Add dependency requirement for `ext-mbstring`.
- Add missing from address to preview images.
- Add type declarations for parameters and return types.
- Add `EmailClient::getInboxUrl()` and `EmailClient::getToastUrl()` for clear
  method signatures, requiring an instance of `SubjectPreview` as the param.
- Add `with*` methods to `SubjectPreview` to create a modified new instance.
### Changed
- **BC break**: Namespace changed to `Phlib\LitmusSubjectPreview`.
- **BC break**: `SubjectPreview` is immutable, requiring constructor parameters.
- **BC break**: `EmailClient::hasInstance()` renamed to `EmailClient::create()` 
  to make it clearer that it returns a new instance, not a singleton.
- Change licence to LGPLv3 to match other Phlib projects.
- Unknown email client throws `DomainException` in
  `EmailClient::getInstance()` instead of generic `Exception`.
- `EmailClient::setHasToast()` returns self instead of the given value, to
  match fluent behaviour of other setters.
- Strings for subject/body/sender are multi-byte safe when truncated.
  Newlines are removed.
- `EmailClient` which does not support toast returns an empty array for 
  `getToastSize()` instead of `null`.
### Removed
- **BC break**: Removed support for PHP versions <= 7.2 as they are no longer
  [actively supported](https://php.net/supported-versions.php)
  by the PHP project.
- **BC break**: Removed `SubjectPreview::getEndpoint()`.
- **BC break**: Removed `SubjectPreview::getEmailClient()`.
  Use `EmailClient::getInstance()`.
- **BC break**: Removed `EmailClient::setSubjectPreview()`.
- **BC break**: Removed `EmailClient::getUrl()`.
  Use `getInboxUrl()` or `getToastUrl()`.
- **BC break**: Removed setters from `SubjectPreview` to give an immutable 
  object using constructor params. Modify an existing subject using `with*` 
  methods.
- **BC break**: Removed setters from `EmailClient` to give an immutable object
  from pre-defined values for each client.

## [2.0.1] - 2021-04-03

- Import subject line preview elements from
  [`phlib/litmus-reseller-sdk`](https://github.com/phlib/litmus-reseller-sdk),
  keeping SemVer compatibility with original
  [`yzalis/litmus`](https://packagist.org/packages/yzalis/litmus).
  This version SHOULD NOT be installed alongside v2.x of
  `phlib/litmus-reseller-sdk` as it contains the same classes.
  This version is for implmentations that want to continue to use these
  features without the Reseller SDK, by simply changing the package name.
  The next major version will be fully split from `phlib/litmus-reseller-sdk`.
- Update maintainer info to Phlib.
  Thanks to [blaugueux](https://github.com/blaugueux)!


tests/EmailClientTest.php
<?php

declare(strict_types=1);

namespace Phlib\LitmusSubjectPreview\Test;

use Phlib\LitmusSubjectPreview\SubjectPreview;
use Phlib\LitmusSubjectPreview\EmailClient;
use PHPUnit\Framework\TestCase;

/**
 * @package Phlib\Litmus-Subject-Preview
 */
class EmailClientTest extends TestCase
{
    public function testGetAvailableEmailClients(): void
    {
        $expected = [
            'ol2003',
            'ol2007',
            'ol2010',
            'hotmail',
            'gmail',
            'yahoo',
        ];

        $actual = EmailClient::getAvailableEmailClients();

        static::assertCount(count($expected), $actual);

        foreach ($expected as $emailClientSlug) {
            static::assertContains($emailClientSlug, $actual);
        }
    }

    public function testGetInstanceInvalid(): void
    {
        $this->expectException(\DomainException::class);
        $this->expectExceptionMessage('does not exist');

        EmailClient::create(sha1(uniqid()));
    }

    /**
     * @dataProvider dataEmailClientMeta
     */
    public function testEmailClientMeta(
        string $slug,
        string $name,
        bool $hasToast,
        int $width,
        int $height,
        ?int $toastWidth,
        ?int $toastHeight
    ): void {
        $emailClient = EmailClient::create($slug);

        static::assertEquals($slug, $emailClient->getSlug());
        static::assertEquals($name, $emailClient->getName());
        static::assertSame($hasToast, $emailClient->getHasToast());
        $size = $emailClient->getGlobalSize();
        static::assertEquals($width, $size['width']);
        static::assertEquals($height, $size['height']);
        $size = $emailClient->getToastSize();
        if ($toastWidth === null) {
            static::assertEmpty($size);
        } else {
            static::assertEquals($toastWidth, $size['width']);
            static::assertEquals($toastHeight, $size['height']);
        }
    }

    public function dataEmailClientMeta(): array
    {
        return [
            'ol2003' => [
                'slug' => 'ol2003',
                'name' => 'Outlook 2003',
                'hasToast' => true,
                'width' => 841,
                'height' => 128,
                'toastWidth' => 329,
                'toastHeight' => 74,
            ],
            'ol2007' => [
                'slug' => 'ol2007',
                'name' => 'Outlook 2007',
                'hasToast' => true,
                'width' => 662,
                'height' => 169,
                'toastWidth' => 329,
                'toastHeight' => 74,
            ],
            'ol2010' => [
                'slug' => 'ol2010',
                'name' => 'Outlook 2010',
                'hasToast' => true,
                'width' => 579,
                'height' => 128,
                'toastWidth' => 329,
                'toastHeight' => 74,
            ],
            'hotmail' => [
                'slug' => 'hotmail',
                'name' => 'Hotmail',
                'hasToast' => false,
                'width' => 687,
                'height' => 110,
                'toastWidth' => null,
                'toastHeight' => null,
            ],
            'gmail' => [
                'slug' => 'gmail',
                'name' => 'Gmail',
                'hasToast' => false,
                'width' => 803,
                'height' => 83,
                'toastWidth' => null,
                'toastHeight' => null,
            ],
            'yahoo' => [
                'slug' => 'yahoo',
                'name' => 'Yahoo',
                'hasToast' => false,
                'width' => 601,
                'height' => 104,
                'toastWidth' => null,
                'toastHeight' => null,
            ],
        ];
    }

    public function testUrlBase(): void
    {
        $subject = sha1(uniqid());
        $body = sha1(uniqid());
        $sender = sha1(uniqid());

        $subjectPreview = new SubjectPreview(
            $subject,
            $body,
            $sender,
        );

        $uri = EmailClient::create('ol2003')->getInboxUrl($subjectPreview);

        static::assertStringStartsWith('https://allclients.litmus.com/s/?c=ol2003', $uri);

        $query = parse_url($uri, PHP_URL_QUERY);
        $params = [];
        parse_str($query, $params);

        static::assertArrayHasKey('s', $params);
        static::assertEquals($subject, $params['s']);

        static::assertArrayHasKey('p', $params);
        static::assertEquals($body, $params['p']);

        static::assertArrayHasKey('f', $params);
        static::assertEquals($sender, $params['f']);

        static::assertArrayHasKey('rnd', $params);
        static::assertIsNumeric($params['rnd']);
    }

    public function testUrlInbox(): void
    {
        $subjectPreview = new SubjectPreview(
            sha1(uniqid()),
            sha1(uniqid()),
            sha1(uniqid()),
        );

        $uri = EmailClient::create('ol2003')->getInboxUrl($subjectPreview);

        $query = parse_url($uri, PHP_URL_QUERY);
        $params = [];
        parse_str($query, $params);

        static::assertArrayHasKey('t', $params);
        static::assertEquals('subject', $params['t']);
    }

    public function testUrlToast(): void
    {
        $subjectPreview = new SubjectPreview(
            sha1(uniqid()),
            sha1(uniqid()),
            sha1(uniqid()),
        );

        $uri = EmailClient::create('ol2003')->getToastUrl($subjectPreview);

        $query = parse_url($uri, PHP_URL_QUERY);
        $params = [];
        parse_str($query, $params);

        static::assertArrayHasKey('t', $params);
        static::assertEquals('toast', $params['t']);
    }

    public function testUrlToastNone(): void
    {
        $this->expectException(\DomainException::class);
        $this->expectExceptionMessage('Toast not supported');

        $subjectPreview = new SubjectPreview(
            sha1(uniqid()),
            sha1(uniqid()),
            sha1(uniqid()),
        );

        EmailClient::create('yahoo')->getToastUrl($subjectPreview);
    }
}


Tests/SubjectPreviewTest.php
<?php

namespace Litmus\Tests;

use Litmus\SubjectPreview\SubjectPreview;
use Litmus\SubjectPreview\EmailClient;

/**
 * @package Phlib\Litmus-Subject-Preview
 */
class SubjectPreviewTest extends \PHPUnit_Framework_TestCase
{
    protected $subjectPreview;

    public function setUp()
    {
        $this->subjectPreview = new SubjectPreview();
        $this->subjectPreview
            ->setSubject("abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz")
            ->setBody("zyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfedcba")
            ->setSender("aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz")
        ;
    }

    protected function tearDown()
    {
        $this->subjectPreview = null;
    }

    public function testSubjectPreview()
    {
        $this->assertEquals(100, strlen($this->subjectPreview->getSubject()));
        $this->assertEquals(100, strlen($this->subjectPreview->getBody()));
        $this->assertEquals(50, strlen($this->subjectPreview->getSender()));
    }

    public function testEmailClient()
    {
        $this->assertCount(6, EmailClient::getAvailableEmailClients());

        foreach (EmailClient::getAvailableEmailClients() as $emailClientSlug) {
            $this->assertInstanceOf('Litmus\SubjectPreview\EmailClient', $this->subjectPreview->getEmailClient($emailClientSlug));
        }

        $emailClients = EmailClient::getAvailableEmailClients();
        $emailClient = $this->subjectPreview->getEmailClient($emailClients[0]);

        $this->assertEquals('ol2003', $emailClient->getSlug());
        $this->assertEquals('Outlook 2003', $emailClient->getName());
        $this->assertTrue($emailClient->getHasToast());
        $size = $emailClient->getGlobalSize();
        $this->assertEquals(128, $size['height']);
        $this->assertEquals(841, $size['width']);
        $size = $emailClient->getToastSize();
        $this->assertEquals(74, $size['height']);
        $this->assertEquals(329, $size['width']);
    }

    public function testUrl()
    {
        $this->assertStringStartsWith(
            'https://allclients.litmus.com/s/?c=ol2003' .
            '&s=abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuv' .
            '&p=zyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfedcbazyxwvutsrqponmlkjihgfe' .
            '&f=aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyy' .
            '&t=subject&rnd=',
            $this->subjectPreview->getEmailClient('ol2003')->getUrl()
        );
    }
}


tests/SubjectPreviewTest.php
<?php

declare(strict_types=1);

namespace Phlib\LitmusSubjectPreview\Test;

use Phlib\LitmusSubjectPreview\SubjectPreview;
use PHPUnit\Framework\TestCase;

/**
 * @package Phlib\Litmus-Subject-Preview
 */
class SubjectPreviewTest extends TestCase
{
    /**
     * @dataProvider dataGetSubject
     */
    public function testGetSubject(string $value, string $expected): void
    {
        $subjectPreview = new SubjectPreview(
            $value,
            '',
            '',
        );

        static::assertSame($expected, $subjectPreview->getSubject());
    }

    /**
     * @dataProvider dataGetSubject
     */
    public function testWithSubject(string $value, string $expected): void
    {
        $subject = sha1(uniqid());
        $body = sha1(uniqid());
        $sender = sha1(uniqid());

        $subjectPreview = new SubjectPreview(
            $subject,
            $body,
            $sender,
        );

        $newSubject = $subjectPreview->withSubject($value);

        // Original instance is unmodified
        static::assertSame($subject, $subjectPreview->getSubject());

        // New instance has cleaned new value
        static::assertSame($expected, $newSubject->getSubject());

        // New instance has original values for other properties
        static::assertSame($body, $newSubject->getBody());
        static::assertSame($sender, $newSubject->getSender());
    }

    public function dataGetSubject(): array
    {
        $baseString = sha1(uniqid()); // 40 chars
        $multibyte = '€'; // Euro \u{20AC}
        return [
            'shortSubject' => [
                str_repeat($baseString, 2),
                str_repeat($baseString, 2),
            ],
            'shortSenderNewline' => [
                $baseString . "\r\n" . $baseString,
                $baseString . ' ' . $baseString, // Replace newline
            ],
            'longSubject' => [
                str_repeat($baseString, 5),
                substr(str_repeat($baseString, 3), 0, 100), // Truncate to 100 chars
            ],
            'longSubjectMb' => [
                str_repeat($baseString, 2) . $multibyte . str_repeat($baseString, 2),
                str_repeat($baseString, 2) . $multibyte . substr($baseString, 0, 19), // MB-safe truncate to 100 chars
            ],
            'longSenderNewline' => [
                str_repeat($baseString, 2) . "\r\n" . str_repeat($baseString, 2),
                str_repeat($baseString, 2) . ' ' . substr($baseString, 0, 19), // Replace newline, truncate to 100 chars
            ],
        ];
    }

    /**
     * @dataProvider dataGetBody
     */
    public function testGetBody(string $value, string $expected): void
    {
        $subjectPreview = new SubjectPreview(
            '',
            $value,
            '',
        );

        static::assertSame($expected, $subjectPreview->getBody());
    }

    /**
     * @dataProvider dataGetBody
     */
    public function testWithBody(string $value, string $expected): void
    {
        $subject = sha1(uniqid());
        $body = sha1(uniqid());
        $sender = sha1(uniqid());

        $subjectPreview = new SubjectPreview(
            $subject,
            $body,
            $sender,
        );

        $newSubject = $subjectPreview->withBody($value);

        // Original instance is unmodified
        static::assertSame($body, $subjectPreview->getBody());

        // New instance has cleaned new value
        static::assertSame($expected, $newSubject->getBody());

        // New instance has original values for other properties
        static::assertSame($subject, $newSubject->getSubject());
        static::assertSame($sender, $newSubject->getSender());
    }

    public function dataGetBody(): array
    {
        $baseString = sha1(uniqid()); // 40 chars
        $multibyte = '€'; // Euro \u{20AC}
        return [
            'shortBody' => [
                str_repeat($baseString, 2),
                str_repeat($baseString, 2),
            ],
            'shortSenderNewline' => [
                $baseString . "\r\n" . $baseString,
                $baseString . ' ' . $baseString, // Replace newline
            ],
            'longBody' => [
                str_repeat($baseString, 5),
                substr(str_repeat($baseString, 3), 0, 100), // Truncate to 100 chars
            ],
            'longSubjectMb' => [
                str_repeat($baseString, 2) . $multibyte . str_repeat($baseString, 2),
                str_repeat($baseString, 2) . $multibyte . substr($baseString, 0, 19), // MB-safe truncate to 100 chars
            ],
            'longSenderNewline' => [
                str_repeat($baseString, 2) . "\r\n" . str_repeat($baseString, 2),
                str_repeat($baseString, 2) . ' ' . substr($baseString, 0, 19), // Replace newline, truncate to 100 chars
            ],
        ];
    }

    /**
     * @dataProvider dataGetSender
     */
    public function testGetSender(string $value, string $expected): void
    {
        $subjectPreview = new SubjectPreview(
            '',
            '',
            $value,
        );

        static::assertSame($expected, $subjectPreview->getSender());
    }

    /**
     * @dataProvider dataGetSender
     */
    public function testWithSender(string $value, string $expected): void
    {
        $subject = sha1(uniqid());
        $body = sha1(uniqid());
        $sender = sha1(uniqid());

        $subjectPreview = new SubjectPreview(
            $subject,
            $body,
            $sender,
        );

        $newSubject = $subjectPreview->withSender($value);

        // Original instance is unmodified
        static::assertSame($sender, $subjectPreview->getSender());

        // New instance has cleaned new value
        static::assertSame($expected, $newSubject->getSender());

        // New instance has original values for other properties
        static::assertSame($subject, $newSubject->getSubject());
        static::assertSame($body, $newSubject->getBody());
    }

    public function dataGetSender(): array
    {
        $baseString = sha1(uniqid()); // 40 chars
        $multibyte = '€'; // Euro \u{20AC}
        return [
            'shortSender' => [
                $baseString,
                $baseString,
            ],
            'shortSenderNewline' => [
                $baseString . "\r\n",
                $baseString . ' ', // Replace newline
            ],
            'longSender' => [
                str_repeat($baseString, 5),
                substr(str_repeat($baseString, 3), 0, 50), // Truncate to 50 chars
            ],
            'longSubjectMb' => [
                $baseString . $multibyte . $baseString,
                $baseString . $multibyte . substr($baseString, 0, 9), // MB-safe truncate to 50 chars
            ],
            'longSenderNewline' => [
                $baseString . "\r\n" . $baseString,
                $baseString . ' ' . substr($baseString, 0, 9), // Replace newline, truncate to 50 chars
            ],
        ];
    }
}


example/example.php
<?php

declare(strict_types=1);

require __DIR__ . '/../vendor/autoload.php';

use Phlib\LitmusSubjectPreview\EmailClient;
use Phlib\LitmusSubjectPreview\SubjectPreview;

// Build subject details
$subjectPreview = new SubjectPreview(
    'Lorem ipsum',
    'Lorem ipsum dolor sit amet',
    'Example sender <sender@example.com>',
);

// Fluent access to single client's preview URI
$previewUri = EmailClient::create('ol2003')->getInboxUrl($subjectPreview);
echo '<pre>' . $previewUri . '</pre>';

echo "\n\n";

// Output all available images
foreach (EmailClient::getAvailableEmailClients() as $clientName) {
    $emailClient = EmailClient::create($clientName);
    echo "<h1>{$emailClient->getName()} ({$emailClient->getSlug()})</h1>\n";

    $subjectSize = $emailClient->getGlobalSize();
    echo <<<HTML
<img src="{$emailClient->getInboxUrl($subjectPreview)}"
    alt="{$emailClient->getSlug()} subject"
    width="{$subjectSize['width']}"
    height="{$subjectSize['height']}"
    />

HTML;

    if ($emailClient->getHasToast()) {
        $toastSize = $emailClient->getToastSize();
        echo <<<HTML
<img src="{$emailClient->getToastUrl($subjectPreview)}"
    alt="{$emailClient->getSlug()} toast"
    width="{$toastSize['width']}"
    height="{$toastSize['height']}"
    />

HTML;
    }

    echo "\n\n";
}


src/SubjectPreview.php
<?php

declare(strict_types=1);

namespace Phlib\LitmusSubjectPreview;

/**
 * Filtered values for generating email's subject pictures
 *
 * @package Phlib\Litmus-Subject-Preview
 */
class SubjectPreview
{
    /**
     * @var string The email subject
     */
    private $subject;

    /**
     * @var string The email body
     */
    private $body;

    /**
     * @var string The email sender name
     */
    private $sender;

    public function __construct(string $subject, string $body, string $sender)
    {
        $this->subject = $this->clean($subject, 100);
        $this->body = $this->clean($body, 100);
        $this->sender = $this->clean($sender, 50);
    }

    public function withSubject(string $subject): self
    {
        return new static($subject, $this->body, $this->sender);
    }

    public function withBody(string $body): self
    {
        return new static($this->subject, $body, $this->sender);
    }

    public function withSender(string $sender): self
    {
        return new static($this->subject, $this->body, $sender);
    }

    public function getSubject(): string
    {
        return $this->subject;
    }

    public function getBody(): string
    {
        return $this->body;
    }

    public function getSender(): string
    {
        return $this->sender;
    }

    private function clean(string $text, int $substrSize): string
    {
        // strip out newline characters
        $text = str_replace(["\r\n", "\r", "\n"], ' ', $text);

        $text = mb_substr($text, 0, $substrSize);

        // Litmus-specific substitutions
        $text = str_replace(['&', '+', '#'], ['$AMP;', '$PLUS;', '$HASH;'], $text);

        return $text;
    }
}


src/EmailClient.php
<?php

declare(strict_types=1);

namespace Phlib\LitmusSubjectPreview;

/**
 * @package Phlib\Litmus-Subject-Preview
 */
class EmailClient
{
    /**
     * @var string
     */
    private $baseUri = 'https://allclients.litmus.com/s/';

    /**
     * @var string The email client name
     */
    private $name;

    /**
     * @var string The email client identifier
     */
    private $slug;

    /**
     * @var bool Define if the email client has a toast view
     */
    private $hasToast;

    /**
     * @var array The width and height of global subject preview
     */
    private $globalSize;

    /**
     * @var array The width and height of toast subject preview
     */
    private $toastSize;

    /**
     * @var array $clientsDatas List of mail client datas
     */
    private static $clientsDatas = [
        'ol2003' => [
            'name' => 'Outlook 2003',
            'hasToast' => true,
            'globalSize' => ['width' => 841, 'height' => 128],
            'toastSize' => ['width' => 329, 'height' => 74],
        ],
        'ol2007' => [
            'name' => 'Outlook 2007',
            'hasToast' => true,
            'globalSize' => ['width' => 662, 'height' => 169],
            'toastSize' => ['width' => 329, 'height' => 74],
        ],
        'ol2010' => [
            'name' => 'Outlook 2010',
            'hasToast' => true,
            'globalSize' => ['width' => 579, 'height' => 128],
            'toastSize' => ['width' => 329, 'height' => 74],
        ],
        'hotmail' => [
            'name' => 'Hotmail',
            'hasToast' => false,
            'globalSize' => ['width' => 687, 'height' => 110],
        ],
        'gmail' => [
            'name' => 'Gmail',
            'hasToast' => false,
            'globalSize' => ['width' => 803, 'height' => 83],
        ],
        'yahoo' => [
            'name' => 'Yahoo',
            'slug' => 'yahoo',
            'hasToast' => false,
            'globalSize' => ['width' => 601, 'height' => 104],
        ],
    ];

    /**
     * Get available email client slug list
     *
     * @return string[]
     */
    public static function getAvailableEmailClients(): array
    {
        $availableEmailClients = [];
        foreach (self::$clientsDatas as $key => $values) {
            $availableEmailClients[] = $key;
        }

        return $availableEmailClients;
    }

    public static function create(string $slug): EmailClient
    {
        if (!isset(self::$clientsDatas[$slug])) {
            throw new \DomainException(sprintf('The email client "%s" does not exist.', $slug));
        }

        return new EmailClient(
            $slug,
            self::$clientsDatas[$slug]['name'],
            self::$clientsDatas[$slug]['hasToast'],
            self::$clientsDatas[$slug]['globalSize'],
            self::$clientsDatas[$slug]['toastSize'] ?? [],
        );
    }

    public function __construct(
        string $slug,
        string $name,
        bool $hasToast,
        array $globalSize,
        array $toastSize
    ) {
        $this->slug = $slug;
        $this->name = $name;
        $this->hasToast = $hasToast;
        $this->globalSize = $globalSize;
        $this->toastSize = $toastSize;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getSlug(): string
    {
        return $this->slug;
    }

    public function getGlobalSize(): array
    {
        return $this->globalSize;
    }

    public function getToastSize(): array
    {
        return $this->toastSize;
    }

    public function getHasToast(): bool
    {
        return $this->hasToast;
    }

    public function getInboxUrl(SubjectPreview $subject): string
    {
        return $this->getUrl($subject, false);
    }

    public function getToastUrl(SubjectPreview $subject): string
    {
        if (!$this->getHasToast()) {
            throw new \DomainException(sprintf('Toast not supported for "%s"', $this->getSlug()));
        }
        return $this->getUrl($subject, true);
    }

    private function getUrl(SubjectPreview $subject, bool $toast = false): string
    {
        $data = [
            'c' => $this->getSlug(),
            's' => $subject->getSubject(),
            'p' => $subject->getBody(),
            'f' => $subject->getSender(),
            't' => $toast ? 'toast' : 'subject',
            'rnd' => rand(0, 99999)
        ];

        return $this->baseUri . '?' . http_build_query($data);
    }
}
